Network/Socket/Syscall.hs

```diff
@@ -73,20 +72,29 @@ socket :: Family         -- Family Name (usually AF_INET)
        -> ProtocolNumber -- Protocol Number (getProtocolByName to find value)
        -> IO Socket      -- Unconnected Socket
 socket family stype protocol = do
-    c_stype <- packSocketTypeOrThrow "socket" stype
+    c_stype <- modifyFlag <$> packSocketTypeOrThrow "socket" stype
+    fd <- throwSocketErrorIfMinus1Retry "Network.Socket.socket" $
+              c_socket (packFamily family) c_stype protocol
+    setNonBlock fd `E.onException` c_close fd
```

Comment:
Won't this result in a `Socket` with a closed file descriptor within its `IORef`?